sympy/core/add.py

```diff
@@ -216,10 +216,25 @@ def flatten(cls, seq):
                         # e.g. 3 + ...
         order_factors = []
 
-        extra = []
+        saw_inf = None  # flag to consult when encountering zoo
+
+        extra, seq = sift(seq, lambda _: isinstance(_, MatrixExpr), binary=True)
```

Comment:
I don't think that a sum could be a mixture of scalars and matrices (or tensors). It should be an error (maybe `ValueError`) if both `extra` and `seq` were nonempty. If `seq` is empty, then there is no need to deal with infinities or `nan`, and `extra` could probably be returned without much further processing; in particular, it would not be necessary to add `nan` to the argument list.

Reply:
> don't add branches in the sympy repo

I guess this happens if a file is modified online and a PR is started. I am not sure how to avoid this other than not starting a pr this way. I think it does this even if I have a URL with smichr/sympy in it since when I edit a file it doesn't have smichr/sympy it has sympy/sympy in the path.


I will reopen and delete this version.

Reply:
> I don't think that a sum could be a mixture of scalars and matrices

This is a separate issue and an inline in `test_matexpr` says,
```
    # Some of these are nonsensical, but we do not raise errors for Add
    # because that breaks algorithms that want to replace matrices with dummy
    # symbols.
```
In this PR I am not focusing on what to do with MatExpr, just trying to faithfully preserve the intent of the code as it is wrt MatExpr and Add.flatten.